GAE/src/org/waterforpeople/mapping/dataexport/RawDataSpreadsheetImporter.java

```diff
@@ -1006,91 +1007,121 @@ protected void invokeUrl(String serverBase, String urlString,
                 urlString, shouldSign, key);
     }
 
-    /*
-     * validate
-     * Called from Clojure code before executeImport()
+    /**
+     * @param sheet
+     * @param isBaseSheet
+     * @param errorMap
+     * @return number of questions columns found
      */
-    @Override
-    public Map<Integer, String> validate(File file) {
-        Map<Integer, String> errorMap = new HashMap<Integer, String>();
+    private int validateSheet(Sheet sheet, boolean isBaseSheet, Map<Integer, String> errorMap) {
+        //Each sheet should have a "Metadata" header, and on the next row, column headers
+        //Verify that this is a 2017-style report w group headers and rqg's on separate sheets
+        if (!safeCellCompare(sheet, 0, 0, METADATA_LABEL)) {
+            errorMap.put(0, "First header cell on each sheet must contain '" + METADATA_LABEL + "'");
+            return 0;
+        }
 
-        try {
-            Sheet sheet = getDataSheet(file);
+        int questionCount = 0;
+        final int headerRowIndex = 1; //Always, now
+        Row headerRow = sheet.getRow(headerRowIndex);
+        int firstQuestionColumnIndex = -1;
+        int lastNonemptyHeaderColumnIndex = 0;
+
+        for (Cell cell : headerRow) {
+            String cellValue = cell.getStringCellValue();
+            // if encountering a null cell make sure its only due to phantom cells at the end of
+            // the row. If null or empty cell occurs in middle of header row report an error
+            if (isEmptyCell(cell) && nonEmptyHeaderCellsAfter(cell)) {
+                errorMap.put(
+                        cell.getColumnIndex(),
+                        String.format(
+                                "Cannot import data from Column %s - \"%s\". Please check and/or fix the header cell",
+                                CellReference.convertNumToColString(cell.getColumnIndex()),
+                                cellValue));
 
-            //Find out if this is a 2017-style report w group headers and rqg's on separate sheets
-            boolean splitSheets = safeCellCompare(sheet, 0, 0, METADATA_LABEL);
-            if (!splitSheets) {
-                errorMap.put(0, "First header cell must contain '" + METADATA_LABEL + "'");
-                return errorMap;
+                break;
             }
 
-            int headerRowIndex = splitSheets ? 1 : 0;
+            if (!isEmptyCell(cell)) {
+                lastNonemptyHeaderColumnIndex = cell.getColumnIndex();
+            }
+            if (cellValue.matches("[0-9]+\\|.+")) {
```

Comment:
I saw you extracted the regex pattern for the sheet, maybe we can extract this one as well to make it more explicit..?

Reply:
Sure